### PR TITLE
Add distributed graph info for distributed training

### PR DESCRIPTION
**Pull request by ZhengHongming888**
This code belongs to the part of the whole distributed training for PyG.

This PR originally designed for the DistGraph class and now merged with LocalGraphStore -

1. Add partition info into LocalGraphStore like num_partition, partition_idx, node_pb (node_partitionbook), edge_pb, partition_meta
2. Two api functions to return the partition idx based on the global ids (node/edge ids) are included. 

Now we combined the local graph and dist graph properties into one GraphStore. later on we will change the class name from LocalGraphStore into PartitionGraphStore with another PR. 

Any comments please let us know. thanks.

### CODE REVIEW

File: torch_geometric/distributed/local_graph_store.py [hunk 1]
```diff
@@ -18,6 +18,13 @@ def __init__(self):
         self._edge_attr: Dict[Tuple, EdgeAttr] = {}
         self._edge_id: Dict[Tuple, Tensor] = {}
 
+        # for partition info related to graph
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:21]: Would recommend documenting `node_pb`, `edge_pb`, and `meta`. Meta is a bit concerning since it seems to be a Dict[Any, Any] with no additional information.

File: torch_geometric/distributed/local_graph_store.py [hunk 2]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+
+    def set_edge_pb(
+            self, edge_pb: Union[torch.Tensor, Dict[NodeType,
+                                                    torch.Tensor]]) -> bool:
+        self.edge_pb = edge_pb
+        return True
+
+    def set_partition_meta(self, partition_meta: Dict) -> bool:
+        self.meta = partition_meta
+        return True
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:85]: * Why do these functions return `bool`?
* We should either define setters and have private instance variables or just use the public instance variables. No need to have setters on public instance variables.

File: torch_geometric/distributed/local_graph_store.py [hunk 3]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+
+    def get_partition_ids_from_nids(self, ids: torch.Tensor,
+                                    ntype: Optional[NodeType] = None):
+        # Get the local partition ids of node ids with a specific node type.
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:89]: nit, move this to a function-level doc, with `r"""<doc>"""`

File: torch_geometric/distributed/local_graph_store.py [hunk 4]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+                                                    torch.Tensor]]) -> bool:
+        self.edge_pb = edge_pb
+        return True
+
+    def set_partition_meta(self, partition_meta: Dict) -> bool:
+        self.meta = partition_meta
+        return True
+
+    def get_partition_ids_from_nids(self, ids: torch.Tensor,
+                                    ntype: Optional[NodeType] = None):
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:88]:
```suggestion
    def get_partition_ids_from_nids(self, ids: torch.Tensor,
                                    node_type: Optional[NodeType] = None):
```
nit

File: torch_geometric/distributed/local_graph_store.py [hunk 5]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+        # Get the local partition ids of node ids with a specific node type.
+        if (self.meta["is_hetero"]):
+            assert ntype is not None
+            return self.node_pb[ntype][ids]
+        return self.node_pb[ids]
+
+    def get_partition_ids_from_eids(self, eids: torch.Tensor,
+                                    etype: Optional[EdgeType] = None):
+        # Get the partition ids of edge ids with a specific edge type.
+        if (self.meta["is_hetero"]):
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:98]:
```suggestion
        if self.meta["is_hetero"]:
```

File: torch_geometric/distributed/local_graph_store.py [hunk 6]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+        return True
+
+    def set_partition_meta(self, partition_meta: Dict) -> bool:
+        self.meta = partition_meta
+        return True
+
+    def get_partition_ids_from_nids(self, ids: torch.Tensor,
+                                    ntype: Optional[NodeType] = None):
+        # Get the local partition ids of node ids with a specific node type.
+        if (self.meta["is_hetero"]):
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:90]:
```suggestion
        if self.meta["is_hetero"]:
```

File: torch_geometric/distributed/local_graph_store.py [hunk 7]
```diff
@@ -51,6 +58,48 @@ def _remove_edge_index(self, edge_attr: EdgeAttr) -> bool:
...
+            assert ntype is not None
+            return self.node_pb[ntype][ids]
+        return self.node_pb[ids]
+
+    def get_partition_ids_from_eids(self, eids: torch.Tensor,
+                                    etype: Optional[EdgeType] = None):
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:96]:
```suggestion
                                    edge_type: Optional[EdgeType] = None):
```

**Review by mananshah99 — COMMENTED** [review]
Mostly looks good. Thanks for unifying the data into the `LocalGraphStore` class, I'm optimistic this will be cleaner going forward.

File: torch_geometric/distributed/local_graph_store.py (hunk 7 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:96]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 6 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:90]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 5 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:98]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 4 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:88]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 3 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:89]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 1 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:21]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 2 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:85]: the current style is aligned with _edge_attr and _edge_id and all to be set later by set_xxx(). The bool return is also the same as from the original set_edge_index().

File: torch_geometric/distributed/dist_context.py
Comment (mananshah99) [inline torch_geometric/distributed/dist_context.py:1]: Can we rebase this PR on the other one to avoid this duplicated diff?

File: torch_geometric/distributed/local_graph_store.py (hunk 2 above)
Comment (mananshah99) [reply torch_geometric/distributed/local_graph_store.py:57]: I don't think we need that here. These are just instance variables, and don't need setters and getters as they are not private and these setters/getters do not perform any additional mutation. Would be good to clean that up.

File: torch_geometric/distributed/local_graph_store.py [hunk 8]
```diff
@@ -126,15 +184,18 @@ def from_partition(cls, root: str, pid: int) -> 'LocalGraphStore':
...
             graph_store.put_edge_id(graph_data['edge_id'], **attr)
 
         if meta['is_hetero']:
             for edge_type, data in graph_data.items():
                 attr = dict(edge_type=edge_type, layout='coo',
                             size=data['size'])
-                graph_store.put_edge_index((data['row'], data['col']), **attr)
+                graph_store.put_edge_index(
+                    torch.stack((data['row'], data['col']), dim=0), **attr)
+                #graph_store.put_edge_index((data['row'], data['col']), **attr)
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:198]:
```suggestion
```

File: torch_geometric/distributed/local_graph_store.py [hunk 9]
```diff
@@ -126,15 +184,18 @@ def from_partition(cls, root: str, pid: int) -> 'LocalGraphStore':
 
         if not meta['is_hetero']:
             attr = dict(edge_type=None, layout='coo', size=graph_data['size'])
-            graph_store.put_edge_index((graph_data['row'], graph_data['col']),
-                                       **attr)
+            graph_store.put_edge_index(
+                torch.stack((graph_data['row'], graph_data['col']), dim=0),
+                **attr)
```
Comment (mananshah99) [inline torch_geometric/distributed/local_graph_store.py:189]: Why are we stacking here? put_edge_index expects a tuple. Same below.

**Review by mananshah99 — COMMENTED** [review]
Mostly LGTM, apart from these few comments.

File: torch_geometric/distributed/local_graph_store.py (hunk 9 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:189]: The reason :  edge_index (Tuple[torch.Tensor, torch.Tensor]) defined here can not pass through mp.spawn due to Tuple and used with stack to convert into tensor which can pass through mp.spawn.  We need send PartitionGraphStore/.. into subprocess.

File: torch_geometric/distributed/local_graph_store.py (hunk 8 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:198]: removed

File: torch_geometric/distributed/dist_context.py
Comment (ZhengHongming888) [reply torch_geometric/distributed/dist_context.py:1]: sure will do!

File: torch_geometric/distributed/local_graph_store.py [hunk 10]
```diff
@@ -18,10 +18,38 @@ def __init__(self):
         self._edge_attr: Dict[Tuple, EdgeAttr] = {}
         self._edge_id: Dict[Tuple, Tensor] = {}
 
+        self.num_partitions: int = 1
+        self.partition_idx: int = 0
+        self.node_pb: Union[torch.Tensor, Dict[NodeType, torch.Tensor]] = None
```
Comment (rusty1s) [inline torch_geometric/distributed/local_graph_store.py:23]:
```suggestion
        self.node_pb: Union[Tensor, Dict[NodeType, Tensor]] = None
```

File: torch_geometric/distributed/local_graph_store.py [hunk 11]
```diff
@@ -18,10 +18,38 @@ def __init__(self):
...
+        r""" node_partition_book:  mapping between node ids and partition idx """
```
Comment (rusty1s) [inline torch_geometric/distributed/local_graph_store.py:24]: Move the comment up. Also use `#` instead of `"""`.

File: torch_geometric/distributed/local_graph_store.py [hunk 12]
```diff
@@ -18,10 +18,38 @@ def __init__(self):
...
+        self.edge_pb: Union[torch.Tensor, Dict[EdgeType, torch.Tensor]] = None
+        r""" edge_partition_book:  mapping between edge ids and partition idx """
+        self.meta: Optional[Dict[Any, Any]] = None
+        r""" meta information related to partition and graph store info """
+        self.labels: Union[torch.Tensor, Dict[EdgeType, torch.Tensor]] = None
```
Comment (rusty1s) [inline torch_geometric/distributed/local_graph_store.py:29]: What is `self.labels` and why is it related to edge type information?

File: torch_geometric/distributed/local_graph_store.py (hunk 9 above)
Comment (rusty1s) [reply torch_geometric/distributed/local_graph_store.py:159]: First time I hear that tuples are a problem with `mp.spawn`. Are you sure this info is correct?

Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:159]: sure. i will check again and let my colleague check together. Also in from torch_geometric.testing import get_random_edge_index,  #L22 edge_index = torch.stack([row, col], dim=0) also used stack to convert into tensor instead of EdgeTensorTuple = Tuple[Tensor, Tensor]

File: torch_geometric/distributed/local_graph_store.py (hunk 12 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:29]: normally the dataset will also have the labels y and when we do partition we also can move this part into partition file but also we can put outside. if outside for label we need input this label as argument in NeighborLoader and also for the following sub-funcs like dist_neighbor_sampler, etc

File: torch_geometric/distributed/local_graph_store.py (hunk 10 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:23]: done

File: torch_geometric/distributed/local_graph_store.py (hunk 11 above)
Comment (ZhengHongming888) [reply torch_geometric/distributed/local_graph_store.py:24]: done
